### ISSUE TITLE
Move UrlHelperExtensions.cs from Microsoft.AspNetCore.Mvc.Core to Microsoft.AspNetCore.Mvc.Abstractions

### ISSUE THREAD

**Issue by pushist1y** [aspnet/Mvc#5466]
So assemblies in multi-assembly projects could reference only abstractions.

**Comment by pushist1y** [comment]
Nevermind. It's not registered in services anyway...

**Comment by pushist1y** [comment]
Lurking through sources i've managed to use UrlHelper in side library, so the main point still stands.

Move UrlHelperExtensions.cs, IActionContextAccessor.cs, IUrlHelperFactory.cs to Microsoft.AspNetCore.Mvc.Abstractions

Also consider registering IActionContextAccessor to services in AddMvc() method.

**Comment by Eilon** [comment]
@pushist1y is there a reason that you don't want the libraries to references Mvc.Core?

@rynowak any thoughts on this?

**Comment by pushist1y** [comment]
@Eilon well, there is no _particular_ reason, it's purely architecture issue. Following loose coupling pattern you should not reference implementations, you should reference abstractions. And provided you already have an abstractions library there is no good reason in having Mvc.Core in the dependencies list when you can restrict it to abstractions.

**Comment by Eilon** [comment]
I talked to @rynowak today and we feel that the current location of these types is acceptable. The idea of the Abstractions package is to have exactly that - just abstractions (and some of the necessary *context* types), and as little actual functionality as possible. These particular types are mostly associated with the particular *implementation* of MVC that we built and thus belong at a different layer of the system.

While there are no doubt some places where we didn't get things exactly right, we feel that the value of moving these types is minimal compared to the cost.
